sam/bamWrite.go

```diff
@@ -0,0 +1,335 @@
+package sam
+
+import (
+	"bytes"
+	"encoding/hex"
+	"github.com/vertgenlab/gonomics/bgzf"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/exception"
+	"io"
+	"log"
+	"strconv"
+	"strings"
+)
+
+// BamWriter wraps a bgzf.Writer and provides functions to write
+// a Sam struct to a binary Bam file, including WriteToBamFileHandle.
+type BamWriter struct {
+	bgzf.Writer
+	buf       bytes.Buffer // for storage between records
+	recordBuf bytes.Buffer // for use within 1 record
+	refMap    map[string]int
+	u32       [4]byte // bytes for writing little endian
+}
+
+const (
+	nul byte = '\u0000'
+)
+
+// NewBamWriter creates a new bam writer to the input io.Writer.
+// The magic bam bytes and header are immediately written to the BamWriter.
+func NewBamWriter(w io.Writer, h Header) *BamWriter {
+	var bw BamWriter
+	bw.Writer = bgzf.NewWriter(w)
+	bw.buf.WriteString(magicBam)
+
+	// write len of header text
+	headerText := strings.Join(h.Text, "\n") + "\n"
+	le.PutUint32(bw.u32[:], uint32(len(headerText)))
+	bw.buf.Write(bw.u32[:])
+
+	// write header text
+	bw.buf.WriteString(headerText)
+
+	// write number of ref sequences
+	le.PutUint32(bw.u32[:], uint32(len(h.Chroms)))
+	bw.buf.Write(bw.u32[:])
+
+	// write reference information
+	bw.refMap = make(map[string]int)
+	for i, ref := range h.Chroms {
+		// save order to refMap
+		bw.refMap[ref.Name] = i
+
+		// write len of ref name + 1 for nul byte
+		le.PutUint32(bw.u32[:], uint32(len(ref.Name)+1))
+		bw.buf.Write(bw.u32[:])
+
+		// write ref name + nul byte
+		bw.buf.WriteString(ref.Name)
+		bw.buf.WriteByte(nul)
+
+		// write ref len
+		le.PutUint32(bw.u32[:], uint32(ref.Size))
+		bw.buf.Write(bw.u32[:])
+	}
+	return &bw
+}
+
+// Close writes any data remaining in the buffer and closes
+// the underlying bgzf.Writer.
+func (bw *BamWriter) Close() error {
+	bytesRemain := bw.buf.Len()
+	n, err := bw.Write(bw.buf.Bytes())
+	if n != bytesRemain || err != nil {
+		log.Panic(err)
+	}
+	return bw.Writer.Close()
+}
+
+// WriteToBamFileHandle writes a single Sam struct to a bam file.
+func WriteToBamFileHandle(bw *BamWriter, s Sam, bin uint16) {
+	bw.recordBuf.Reset()
+
+	// refID
+	idx, ok := bw.refMap[s.RName]
+	if !ok {
+		idx = -1
```

Comment:
Does this mean that the read is mapping to a chromosome/scaffold that is not in the header?  I would have thought that is not allowed / is an error.  I realize some reads may be unmapped, but I am guessing that is something we can explicitly look for.

Reply:
I was using this to catch unmapped reads, but you are right that an explicit check would be more appropriate. Changed.

Reply:
Do we need this check to be above the "if !ok" a few lines before?  I am guessing that "*" will not be present in bw.refMap and that unmapped reads will error on line 87 before they get to this if statement that catches them.  I was think we might have something like: if unmapped, then set idx to -1.  else if !ok, then give an error.

Reply:
Yes, nice catch.